pgjdbc/src/test/java/org/postgresql/test/jdbc4/StreamResultSetTest.java

```diff
@@ -0,0 +1,110 @@
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assume.assumeTrue;
+import static org.postgresql.core.ServerVersion.v10;
+import static org.postgresql.test.TestUtil.createTempTable;
+import static org.postgresql.test.TestUtil.haveMinimumServerVersion;
+
+import org.postgresql.test.jdbc2.BaseTest4;
+
+import org.junit.Test;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class StreamResultSetTest extends BaseTest4 {
+  private static final int GENERATED_ROWS = 1000;
+  private static final String SELECT_SLEEP_BETWEEN_ROWS = "select *,pg_sleep(0.001) from series";
+
+  private PreparedStatement statement;
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    createTempTable(con, "series", "val int");
+    try (Statement s = con.createStatement()) {
+      s.execute("insert into series select * from generate_series(1," + GENERATED_ROWS + ")");
+    }
+    // statement that sleeps one milliscond between each rows
+    statement = con.prepareStatement(SELECT_SLEEP_BETWEEN_ROWS);
+  }
+
+  @Test
+  public void testStreamedResults() throws Exception {
+    readResults();
+    // query second time to make sure the state is correct after streaming
+    readResults();
+  }
+
+  @Test
+  public void statementReExecuteCleansUpResources() throws Exception {
+    ResultSet results = statement.executeQuery();
+
+    // read few first items
+    results.next();
+    results.next();
+
+    // re-execute the same statement -> should close result set
+    // query second time to make sure the connection state is still correct after partial streaming
+    readResults();
+
+    assertThat(results.isClosed(), is(true));
+  }
+
+  @Test
+  public void differentStatementExecuteCleansUpResources() throws Exception {
+    try (PreparedStatement statement2 = con.prepareStatement(SELECT_SLEEP_BETWEEN_ROWS)) {
+      ResultSet results = statement2.executeQuery();
+
+      // read few first items
+      results.next();
+      results.next();
+
+      // re-execute the different statement -> must not close the previous result set
+      // query second time to make sure the connection state is still correct after partial streaming
+      readResults();
+
+      assertThat(results.isClosed(), is(false));
+
+      results.next();
+      results.next();
+    }
+  }
+
+  private void readResults() throws SQLException {
+    // the test query seems to work differently on older servers by not pacing the response rows
+    // the streaming features works nicely though
+    assumeTrue(haveMinimumServerVersion(con, v10));
+
+    try (ResultSet results = statement.executeQuery()) {
+      int count = 0;
+      long startTime = currentTimeMillis();
+      long fistRowTime = 0;
+      while (results.next()) {
+        int value = results.getInt(1);
+        if (fistRowTime == 0) {
+          fistRowTime = currentTimeMillis();
+        }
```

Comment:
Isn't reading firstRowTime and startTime make them always "same" in this situation? ResultSet have to read first row with or without streaming, so coming to this point is really fast, because all waiting for first row is made in executeQuery() method. The only situation when it doesn't happen is when resultSet is empty, but it isn't tested in this situation. I think that startTime should be set before line 91 to really check if data packet with first row was read without waiting for data packet with last row.

Reply:
No it's after the first row has been read. Which could be different. Also `fistRow` is a spelling mistake

Reply:
But `executeQuery()` reads first row right? So if `startTime` and `firstRow` both are read after `executeQuery()` then their times are really close, `next()` doesn't affect reading first row in this situation I think.

Reply:
Fixed the typo in variable and moved the startTime before the executeQuery.

Reply:
I rewrote the streaming test. Now it has less hard coded random values and lots of comments explaining what is being tested. I also verified that the server seems to buffer results into 8kB blocks, so countered that by adding some padding to the rows to force them in multiple packets.